repair/repair.cc

```diff
@@ -1540,3 +1540,75 @@ future<> sync_data_with_repair(seastar::sharded<database>& db,
         });
     });
 }
+
+future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, std::unordered_set<dht::token> tokens) {
+    using inet_address = gms::inet_address;
+    return seastar::async([&db, tm = std::move(tm), tokens = std::move(tokens)] () mutable {
+        auto keyspaces = db.local().get_non_system_keyspaces();
+        rlogger.info("bootstrap_with_repair: started with keyspaces={}", keyspaces);
+        auto myip = utils::fb_utilities::get_broadcast_address();
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                rlogger.info("bootstrap_with_repair: keyspace={} does not exist any more, ignoring it", keyspace_name);
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            dht::token_range_vector desired_ranges = strat.get_pending_address_ranges(tm, tokens, myip);
+
+            //Active ranges
+            auto metadata_clone = tm.clone_only_token_map();
+            auto range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Pending ranges
+            metadata_clone.update_normal_tokens(tokens, myip);
+            auto pending_range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Collects the source that will have its range moved to the new node
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+
+            rlogger.info("bootstrap_with_repair: started with keyspace={}, nr_ranges={}", keyspace_name, desired_ranges.size());
+            for (auto& desired_range : desired_ranges) {
+                for (auto& x : range_addresses) {
+                    const range<dht::token>& src_range = x.first;
+                    seastar::thread::maybe_yield();
+                    if (src_range.contains(desired_range, dht::tri_compare)) {
+                        std::vector<inet_address> old_endpoints(x.second.begin(), x.second.end());
+                        auto it = pending_range_addresses.find(desired_range);
+                        if (it == pending_range_addresses.end()) {
+                            throw std::runtime_error(format("Can not find desired_range = {} in pending_range_addresses", desired_range));
+                        }
+
+                        std::unordered_set<inet_address> new_endpoints(it->second.begin(), it->second.end());
+                        rlogger.debug("bootstrap_with_repair: keyspace={}, range={}, old_endpoints={}, new_endpoints={}",
+                                keyspace_name, desired_range, old_endpoints, new_endpoints);
+                        // Due to CASSANDRA-5953 we can have a higher RF then we have endpoints.
+                        // So we need to be careful to only be strict when endpoints == RF
```

Comment:
The language of this comment line is confusion and leaves me uncertain about what you wanted to say. Was it "Therefore we should only initialize `mandatory_nodes` when `RF == old_endpoints`. This is because if `RF >  old_nodes` (it can never be less than) there isn't going to be any replica that is going to loose this range."? 

BTW, we may want to put an assert checking the above invariant.

Reply:
Yes, correct. We can only have RF > = old_nodes.  If Rf > old_nodes, no one is going to lose this range, so no need to set the mandatory_nodes. 

Do you mean asserting  RF >= old_nodes? The correctness of this is supposed to be guaranteed by the replication_strategy layer. The user of it should trust it. It makes little sense for all the users to do the assert check.

Reply:
I added more comments to clarify.